DATAES-772 - Add after-convert entity callbacks support.

rpuch opened the pull request:
A few notes about the changes.

1. I tried to model after-convert callback method parameters after [Reactive]AfterConvertCallback interfaces in spring-data-mongodb, where they accept the following triple: loaded entity, original document from MongoDB, and collection name. So, for Elasticsearch case, I replaced collection name with `IndexCoodrinates` structure. It is available everywhere where entity materialization is made (with one exception, see item 2).
2. All the search/get/find methods in both reactive and non-reactive ElasticsearchTemplate either accept `IndexCoordinates` explicitly or obtain it from an entity class. The only exception was `searchScrollContinue()` which did not accept `IndexCoordinates` explicitly and did not get it from entity class (although the entity class parameter is available there). The method is protected and deemed for internal use only (according to its javadoc), and its twin method `searchScrollStart()` accepts both `Class<T>` and `IndexCoordinates` as parameters, so I took the liberty to add `IndexCoordinates` parameter to `searchScrollContinue()`. If I'm not right here, please let me know what would be a better solution.
3. In `ElasticsearchTemplate` and `ElasticsearchRestTemplate`, I added a `NotifyingConverter` abstraction on top of the `ElasticsearchConverter`. Its responsibility is to make after-convert callbacks while materializing entities from documents; there are a few ways to do it, so it seemed to be 'too dirty' to just leave them as methods in `AbstractElasticsearchTemplate`. Also, it seems that this wrapping should not be tied to `ElasticsearchConverter` implementation, because otherwise for a user overriding it it would be too easy to break this functionality.
But I'm not 100% satisfied with the result, because now `AbstractElasticsearchTemplate` has 2 converters (`ElasticsearchConverter` and `NotifyingConverter`), this may cause confusion.
If you have any thoughts on this, please let me know.

<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAES).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

sothawo commented:
Thanks for this PR,

I'm not commenting on individual files, but rather on your questions (with some additional comments)

ad 1):

I don't think we should expose the `Document` in any of the convert callbacks. In Spring Data Elasticsearch I see this as an internally used class. For the user to work with Spring Data Elasticsearch the entities should be the way to go. I don't see a use case for exposing the Document. If there is something that is not mapped from the entity to the document or back, then this should be rather be reported as error/feature request. Letting users have too broad hooks to customize the mapping will lead to problems when we later change something there and the custom callback handler logic fails. We had this with the custom `ResponseMapper` implementations in version before 4, and for version 4 we want to encapsulate this more than before. 

To have the index in the callback is ok, as a callback implementation might need this information.

ad 2)

the `searchScrollXXX` methods are internal and are used for the streaming API. `searchScrollStart` needs the entity class to instantiate the returned objects and optionally a different index name toi start the search. `searchScrollContinue` needs the entity class only for instantiating the returned entities. The index name is not needed here, neither from the entity or by an additionally parameter. To get the data from Elasticsearch after a scroll is started, _only_ the scroll id is needed. So nothing needs to be changed here.

ad 3)

I don't like wrapping of the callbacks in a `NotifyingConverter` class which is no converter anyway. If you want to have a place where the triggering of callbacks is handled, then the `NotifyingConverter` should implement the `ElasticsearchConverter` interface and take the original instance as delegate and set this as the template's converter. But `ElasticsearchOperations.getElasticsearchConverter()` still would need to return the original instance, so I don' think that this would be a good solution. I'd prefer to have this code as normal functions in the `AbstractElasticsearchTemplate` class.

Additional question: why did you introduce the additional classes `SearchHitsWrapper`,  `OverridingSearchHits`, `SearchScrollHitsWrapper` and `SearchScrollHitsWrapper`. What is missing from the available interfaces and implementations that these wrappers offer?

Generally I don't like to have lombok in production code. I think it's ok for testing classes, to create data classes and stuff like this, but in the production code it leads for example to debugging problems, when the decompiled code does not match the published docs etc.

rpuch commented:
Thank you for your feedback. I'm working on items 1 and 3.

As for item 2, the reason I added `IndexCoordinates` parameter to `searchScrollContinue()` was because it is needed to make a callback (as the callback method requires `IndexCoordinates` to be passed). If we don't add it as a parameter to `searchScrollContinue()`, what is the alternative way to obtain it? It seems that we cannot just derive it from entity class as it may be overridden in `searchScrollStart()`.

rpuch commented:
> Additional question: why did you introduce the additional classes `SearchHitsWrapper`, `OverridingSearchHits`, `SearchScrollHitsWrapper` and `SearchScrollHitsWrapper`. What is missing from the available interfaces and implementations that these wrappers offer?

Please note the following method two method starts:

```
		<T> SearchHits<T> read(Class<T> type, SearchDocumentResponse searchDocumentResponse, IndexCoordinates index) {

			SearchHits<T> searchHits = converter.read(type, searchDocumentResponse);
```

and

```
		<T> SearchScrollHits<T> readScroll(Class<T> type, SearchDocumentResponse searchDocumentResponse,
				IndexCoordinates index) {

			SearchScrollHits<T> searchHits = converter.readScroll(type, searchDocumentResponse);
```

Both of them obtain an instance of `SearchHits` or `SearchScrollHits` from a document and then need to return an equivalent instance with hits possibly replaced by callbacks.

There is only one implementation of both `SearchHits` and `SearchScrollHits` in spring-data-elasticsearch: `SearchHitsImpl`. So, to construct an instance with 'replaced hits' I need to create an instance of `SearchHitsImpl` which requires `scrollId`. But I don't have a `scrollId` in the first method (because there I only have `SearchHits`, not `SearchScrollHits`).

I could invent the following solutions to this problem:

1. Knowing that currently any `SearchHits` is actually `SearchHitsImpl`, downcast `SearchHits` to `SearchScrollHits`, which gives everything that is needed to reconstruct another `SearchHitsImpl` instance with replaced hits. But downcasting smells pretty bad, I preferred to avoid it.
2. Introduce another concrete implementation of `SearchHits` which is not an implementation of `SearchScrollHits`, probably a superclass of `SearchHitsImpl` and using it in the first method. This is problematic because, first, the best suitable name is already taken by `SearchHitsimpl`, and second, it would probably require to change other code that produces just `SearchHits` (and not `SearchScrollHits`) instances for consistency. If this is the way to go, please let me know, but it seemed to be too invasive.
3. Add implementations of both interfaces that just override the hits they return, hence `OverridingSearchHits` and `OverridingSearchScrollHits`. The `XXXWrapper` superclasses are added in hope that they could be useful for similar purposes; but if you think they are not needed, I can inline them in the `OverridingXXX` classes.

What are your thoughts on this?

sothawo commented:
about the `searchScrollContinue()` method: I was wrong that the index name is taken from the class; it's always the one passed in with the `IndexCoordinates` parameter; so the same can then be passed to the continue function in `searchForStream` to have it for the callback

sothawo commented:
about the `SearchHits`(why can't I reply to a comment here):

You don't need to replace the returned `SearchHits` instance, you nedd to exchange the `T` elements that it contains. 
The `getSearchHits()` method returns a list of `SearchHit` that can be modified, so you could map the `SearchHit` elements of this list to a new list where you construct `SearchHit` objects with the same values of the original but with replaced `T` entities. Then clear the original list and all from the new one.

rpuch commented:
> The `getSearchHits()` method returns a list of `SearchHit` that can be modified, so you could map the `SearchHit` elements of this list to a new list where you construct `SearchHit` objects with the same values of the original but with replaced `T` entities. Then clear the original list and all from the new one.

Thank you for the suggestion. This will require us to assume that the list is always mutable. Right now it is true, but still it is an assumption... Of course, if someone breaks this assumption, tests will fail and demonstrate the problem.

I'm working on the fixes.

rpuch commented:
> The `getSearchHits()` method returns a list of `SearchHit` that can be modified, so you could map the `SearchHit` elements of this list to a new list where you construct `SearchHit` objects with the same values of the original but with replaced `T` entities. Then clear the original list and all from the new one.

It turns out that `getSearchHits()` returns an *unmofidiable* list:

```
	@Override
	public List<SearchHit<T>> getSearchHits() {
		return Collections.unmodifiableList(searchHits);
	}
```

So `SearchHitsImpl` is effectively immutable, hence your idea does not seem to work, unfortunately.

I'd stick with 'overriding' interface implementations. What would you suggest?

mp911de commented:
I second PJ's opinion on not introducing a `NotifyingConverter` as that's not the right abstraction level. Instead, we should look forward introducing a `DocumentCallback` similar to what is done for [MongoDB's `ReadDocumentCallback`](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/MongoTemplate.java#L3123-L3145).

A symptom for lacking such an abstraction is that the converter `read(…)` method gets called in multiple places. Ideally, there is a single call to `converter.read(…)`. A `Document` Callback interface would also solve the issue around a `SearchHits` wrapper because the `SearchHits` objects gets created once with already the entity to use (outcome of the `Document` callback).

A note on the `Document` class: We should leave it in the PR to align with the other modules. A good use-case is that ES documents may contain additional information (especially those that weren't written by Spring Data Elasticsearch). These additional details can be evaluated by the callback to enhance the deserialized document. Typical examples are logging the received document or fetching data from other sources (some kind of relation).

sothawo commented:
Hello Roman,
I just thad a call with Mark, the way we want to go is:
- introduce the `DocumentCallback` like it is in Spring Data MongoDB, so that the callbacks can be executed _before_ the `SearchHit(s)`are created, then these don't need to be changed.
- add the `Document` to the callbacks where it is available in addition to the entity and `IndexCoordinates.

rpuch commented:
Hello guys.

I'm looking at the code now, and I am trying to find the best way to implement this. Let me ask a stupid question or two for a better understanding :)

In the current code, the code that constructs `SearchHits` (and `SearchScrollHits`) lives in `ElasticsearchConverter`'s `read()` and `readScroll()` methods. In the new design suggested by Mark, this code will need to get a `DocumentCallback` and use it for conversion of each individual entity (and entity callbacks). So either `ElasticsearchConverter.read()` (and `readScroll()`) need to get a new parameter of type `DocumentCallback`, or this code needs to be moved elsewhere (and these two methods will probably need to be removed from `ElasticsearchConverter`).

To me, it seems that the preferrable way is the second option, but I prefer to ask for your opinions. It could happen, I just misunderstood the idea completely.

mp911de commented:
Thanks for bringing this up.

Until now, creation of `SearchHit`/`SearchHits` in the converter was fine and did not impose any constraints on what we tried to achieve. Going forward with result mapping of converted objects we're required to create `SearchHit`/`SearchHits` on a different level. In MongoDB we use a specific `GeoNearResultDocumentCallback` that creates `GeoResult`.

It would make sense to follow a similar design in which `SearchHit`/`SearchHits` are constructed on a level that is closer to the template API than the converter API.

Looking on the PR and the current state, especially in hindsight of the next release candidate, it would make sense to introduce the change to `SearchHit` creation/`DocumentCallback` first, ideally in a separate PR as we already feel that this PR has grown too much.

In any case, that is a good exercise as it helps to shape the API.

That being said, we should move both `read(…)` methods from `ElasticsearchConverter` that work with `SearchDocumentResponse`/`SearchDocument` to a utility similar to `EntityOperations`. We would also need to put `getHighlightsAndRemapFieldNames` there and in the same PR we would get the `DocumentCallback`. Once these changes are in place, we can continue with this PR.

Does that make sense?

sothawo commented:
a separate PR for this issue or do we create a new issue for that?

mp911de commented:
A separate PR and issue to be more precise.

sothawo commented:
@rpuch do you need any support on this?

rpuch commented:
I've created a new issue in Jira: https://jira.spring.io/browse/DATAES-786
Let's move the discussion there.

rpuch commented:
I've rebased the current PR to the current master (the master having a fix for DATAES-786).

I've removed lombok usage from production code.

I've also refactored tests a bit: tests for both non-reactive template implementations are identical and only differ in initialization, so I extracted the common base (with all 32 tests) and left the initialization in 2 test suite classes.

Could you please take a look?

sothawo commented on src/test/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplateCallbackTests.java:
```diff
@@ -0,0 +1,568 @@
+	@Spy private ValueCapturingAfterSaveCallback afterSaveCallback = new ValueCapturingAfterSaveCallback();
+	@Spy private ValueCapturingAfterConvertCallback afterConvertCallback = new ValueCapturingAfterConvertCallback();
+
+	protected final void initTemplate(AbstractElasticsearchTemplate templateToInit) {
+		templateToInit.setEntityCallbacks(EntityCallbacks.create(afterSaveCallback, afterConvertCallback));
+	}
+
```
```suggestion
	protected final void initTemplate(AbstractElasticsearchTemplate template) {
	        this.template = template;
		template.setEntityCallbacks(EntityCallbacks.create(afterSaveCallback, afterConvertCallback));
	}
	
	the you can remove the assignment to `template` in the derived classes - see my comments there

```

rpuch replied:
Done

sothawo commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTransportTemplateCallbackTests.java:
```diff
@@ -68,10 +62,24 @@
 	@BeforeEach
+	@SuppressWarnings("deprecation") // we know what we are testing
 	public void setUp() {
 		template = new ElasticsearchTemplate(client);
+		initTemplate(template);
```
```suggestion
		initTemplate(new ElasticsearchTemplate(client));
```

with the change in the base class.

rpuch replied:
Fixed

sothawo commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplateCallbackTests.java:
```diff
@@ -15,258 +15,86 @@
+	@SuppressWarnings("deprecation") // we know what we test
 	@BeforeEach
 	public void setUp() throws Exception {
 		template = new ElasticsearchRestTemplate(client);
+		initTemplate(template);
```
```suggestion
		initTemplate(new ElasticsearchRestTemplate(client));
```

with the change in the base class

rpuch replied:
Fixed

sothawo commented on src/test/java/org/springframework/data/elasticsearch/core/ReactiveElasticsearchTemplateCallbackTests.java:
```diff
@@ -43,13 +49,20 @@
 import org.springframework.lang.Nullable;
 import org.springframework.util.CollectionUtils;
 
+import com.google.common.collect.ImmutableMap;
```
same as in the non-reactive classes

rpuch replied:
Fixed

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/AbstractElasticsearchTemplate.java:
```diff
@@ -550,7 +558,8 @@ public T doWith(@Nullable Document document) {
 			}
 
-			return reader.read(type, document);
+			T materializedEntity = reader.read(type, document);
```
Changed it back to `entity`

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplate.java:
```diff
@@ -172,7 +172,7 @@ public String index(IndexQuery query, IndexCoordinates index) {
 
 		Assert.notNull(index, "index must not be null");
-		Assert.notEmpty(query.getIds(), "No Id define for Query");
+		Assert.notEmpty(query.getIds(), "No Id defined for Query");
```
Fixed

rpuch commented on src/main/java/org/springframework/data/elasticsearch/core/ReactiveElasticsearchTemplate.java:
```diff
@@ -920,8 +931,8 @@ public ReadDocumentCallback(EntityReader<? super T, Document> reader, Class<T> t
 
-			T entity = reader.read(type, document);
-			return Mono.just(entity);
+			T materializedEntity = reader.read(type, document);
```
I wanted to give a reader one more hint about the origin of the entity, but just `entity` is fine. I've renamed it back.

rpuch commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplateCallbackTests.java:
```diff
@@ -15,258 +15,86 @@
-		Person entity2 = new Person("init2", "luke2");
+		doReturn(true).when(getResponse).isExists();
+		doReturn(false).when(getResponse).isSourceEmpty();
+		doReturn(ImmutableMap.of("id", "init", "firstname", "luke")).when(getResponse).getSourceAsMap();
```
Done

rpuch commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchRestTemplateCallbackTests.java:
```diff
@@ -15,258 +15,86 @@
+		doReturn(nSearchHits(2)).when(searchResponse).getHits();
+		doReturn("scroll-id").when(searchResponse).getScrollId();
+		doReturn(new BytesArray(new byte[8])).when(searchHit).getSourceRef();
+		doReturn(ImmutableMap.of("id", "init", "firstname", "luke")).when(searchHit).getSourceAsMap();
```
Done

rpuch commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplateTests.java:
```diff
@@ -2561,7 +2561,7 @@ public void shouldRespectSourceFilterWithScanAndScrollForGivenSearchQuery() {
 
 		// given
-		List<IndexQuery> indexQueries = new ArrayList<>();
+		List<IndexQuery> indexQueries;
```
Done

rpuch commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTemplateTests.java:
```diff
@@ -2609,7 +2609,7 @@ public void shouldSortResultsGivenSortCriteriaWithScanAndScroll() {
 
 		// given
-		List<IndexQuery> indexQueries = new ArrayList<>();
+		List<IndexQuery> indexQueries;
```
Done

rpuch commented on src/test/java/org/springframework/data/elasticsearch/core/ElasticsearchTransportTemplateCallbackTests.java:
```diff
@@ -15,49 +15,43 @@
-import org.springframework.lang.Nullable;
-import org.springframework.util.CollectionUtils;
+
+import com.google.common.collect.ImmutableMap;
```
Done

rpuch commented:
I've fixed the today's batch

sothawo commented:
Thanks a lot for these last two PRs!

rpuch commented:
Thank you for your feedback :)
